std/typecons.d

```diff
@@ -2086,9 +2086,16 @@ if (is(T == class) || is(T == interface) || isAssociativeArray!T)
         U stripped;
     }
 
-    void opAssign(T another) @trusted pure nothrow @nogc
+    void opAssign(T another) pure nothrow @nogc
     {
-        stripped = cast(U) another;
+        // If `T` defines `opCast` we must infer the safety
+        static if (hasMember!(T, "opCast"))
+        {
+            // This will allow the compiler to infer the safety of `T.opCast!U`
+            // without generating any runtime cost
+            if (false) { stripped = cast(U) another; }
+        }
+        () @trusted { stripped = cast(U) another; }();
```

Comment:
What is the point of keeping this lambda? Is there a reason not to just remove `@trusted` from the function?

Reply:
For the case where there's not `opCast` member. In which case casting would be un-`@safe`, but we trust that it's done correctly.

Reply:
I feel that the same should happen for the `opAssign` overloads, that are also `@trusted`, as they are subject to the same issue

If you agree, I'll open a separate PR for those.

Reply:
Yes, we can explore that in a separate PR, but I'm not sure if we can do anything for e.g. this one:

https://github.com/dlang/phobos/blob/dbca8521b5164c2ba95401d9bc784ec653e1a157/std/typecons.d#L2101-L2104